README.md

```diff
@@ -22,13 +61,65 @@ Transaction are applied atomically. Queries can't retrieve partially committed d
+
 # Implementation
 
 ## Data structure
-Data is organized into columns. Each column serving a particular type of data, e.g. state or headers. Column consists of an index and a set of 16 value tables for varying value size.  
+Data is organized into columns. Each column serving a particular type of data, e.g. state or headers. Column consists of an index (up to 255) and a set of 16 value tables for varying value size.  
 
 ### Index
-Index is an is mmap-backed dynamically sized probing hash table. Each entry is a page of 64 8-byte entries, making 512 bytes.  Each 64-bit entry contains 32 bits of value address, 4 bits of value table index and 28 bit value `c` derived from  `k`. `c` is computed by skipping `n` high bits of `k` and taking the next 28 bits.  `k` is 256-bit key that is derived from the original key and is uniformly distributed. `n` is current index bit-size. First `n` bits of `k` map `k` to a page. Entries inside the page are unsorted. Empty entry is denoted with a zero value. Empty database starts with `n` = 16, which allows to put just 240 bits of `k` in the value table. 
+Index is an is mmap-backed dynamically sized probing hash table. Each entry is a page of 64 8-byte entries, making 512 bytes. Each 8-byte entry contains 32 bits of value address, 4 bits of value table index and 28 bit value `c` derived from  `k`. `c` is computed by skipping `n` high bits of `k` and taking the next 28 bits. `k` is 256-bit key that is derived from the original key and is uniformly distributed. `n` is current index bit-size. First `n` bits of `k` map `k` to a page. Entries inside the page are unsorted. Empty entry is denoted with a zero value. Empty database starts with `n` = 16, which allows to put just 240 bits of `k` in the value table. 
+
+
+#### Review Notes
+
+TODO looks outdated: value range is not 32 bit but really the remaining size.
+Leaving 64 - 4 - (6 + n) of hash (initially 38 bit).
+Value adress being n + 6 + 4 as we need to cover all entries of similar size case.
+
+To optimize the page search, I could think of : https://event.cwi.nl/damon2009/DaMoN09-KarySearch.pdf.
+Probably using 16 (8 seems to small like 256 over 64, 1/4 chance to conflict) bit of our k bits at start of page or a first search.
+(then we need to access remaining of page but only at a given entry).
+
+2^16 -> 65k pages ~ 32Megs ix size per column initially.
+
+TODO indicates `n` is called `index_bits` in code.
+Also currently `n` maximum is 64 - 10. (but we would want some at least 16 key length so would say 64 - 26 = 38 bit).
+Even 38 bit seems really enough.
+
+So logic is as much key content as possible to allow early key mismatch detection.
+Starting at 2^16 index * 64 = 4M value (2^22 index that needs to be available among all values categories: 26 bit of adressing).
+Remaining 38 bit of key.
+Considering the key content should at least key 16 bit, then max size for n becomes (64 - 10 - 16) = 38 bit (and 48 bit value address).
```

Comment:
These are all valid points. As `n` increases, so does the side and of the address, and the size of  `c` decreases. Eventually leading to more probing of the value table. Although the index will need to be really huge before this becomes an issue.